src/pages/index.js

```diff
@@ -1,68 +1,204 @@
+    openedModal !== null &&
+    openedModal.querySelector(".modal__container") !== null &&
+    !openedModal.querySelector(".modal__container").contains(event.target)
+  ) {
+    closeModal(openedModal);
+  }
+};
+
+const isInitialCardsSynced = () => {
+  return localStorage.getItem("initialCardsSynced") === "true";
+};
 
+const setInitialCardsSynced = () => {
+  localStorage.setItem("initialCardsSynced", "true");
+};
+
+// Function to handle "Esc" key press
+const handleEscKeyPress = (event) => {
+  if (event.key === "Escape") {
+    const openedModal = document.querySelector(".modal_opened");
+    if (openedModal) {
+      closeModal(openedModal);
+    }
+  }
+};
+
+// Add event listeners for clicks and key presses
+document.addEventListener("mousedown", handleOutsideClick);
+document.addEventListener("keydown", handleEscKeyPress);
+
+// Instances
 const handleImageClick = (link, alt, name) => {
   imagePopup.open({ link, alt, name });
 };
 
+const handleDeleteClick = (cardInstance, cardId) => {
+  const confirmDeleteModal = document.querySelector("#modal-confirm-delete");
+  const confirmDeleteButton = confirmDeleteModal.querySelector("#confirm-delete-button");
+
+  confirmDeleteModal.classList.add("modal_opened");
+
+  confirmDeleteButton.onclick = () => {
+    api
+      .deleteCard(cardId)
+      .then(() => {
+        cardInstance.removeCard();
+        closeModal(confirmDeleteModal);
+      })
+      .catch((err) => console.error(err));
+  };
+
+  const closeButton = confirmDeleteModal.querySelector(".modal__close-button");
+  closeButton.onclick = () => closeModal(confirmDeleteModal);
+};
+
+const handleLikeClick = (cardId, isLiked) => {
+  if (isLiked) {
+    return api.likeCard(cardId).then((data) => {
+      return { ...data, isLiked: true };
+    });
+  } else {
+    return api.dislikeCard(cardId).then((data) => {
+      return { ...data, isLiked: false };
+    });
+  }
+};
+
 const userInfo = new UserInfo({
   nameSelector: ".profile__name",
   jobSelector: ".profile__description",
+  avatarSelector: "#profile-image",
 });
 
 const renderCard = (item) => {
-  const card = new Card(item, "#card-template", handleImageClick);
+  const card = new Card(
+    {
+      name: item.name,
+      link: item.link,
+      alt: item.alt,
+      _id: item._id,
+      likes: item.likes,
+      isLiked: item.isLiked,
+    },
+    "#card-template",
+    handleImageClick,
+    handleDeleteClick,
+    handleLikeClick,
+    userInfo
+  );
   const cardElement = card.getView();
   cardList.addItem(cardElement);
 };
 
 const cardList = new Section(
   {
-    items: initialCards.reverse(),
+    items: [],
     renderer: renderCard,
   },
   ".cards__list"
 );
 
-cardList.renderItems();
+// Extend PopupWithForm for loading state handling
+class CustomPopupWithForm extends PopupWithForm {
+  setLoadingState(isLoading) {
+    const saveButton = this._popup.querySelector(".modal__button");
+
+    if (isLoading) {
+      saveButton.textContent = "Saving...";
```

Comment:
**Could be improved**

It’s better to make method `renderLoading` inside class `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
	// here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```